python/ray/tune/schedulers/pb2.py

```diff
@@ -0,0 +1,341 @@
+                 mode: Optional[str] = None,
+                 perturbation_interval: float = 60.0,
+                 hyperparam_bounds: Dict = None,
+                 quantile_fraction: float = 0.25,
+                 log_config: bool = True,
+                 require_attrs: bool = True,
+                 synch: bool = False):
+
+        if GPy is None:
+            raise RuntimeError("Need to install GPy to use PB2")
+
+        resample_probability = 0
+        custom_explore_fn = None
+        hyperparam_mutations = None
+
+        super(PopulationBasedTraining, self).__init__(
+            self, time_attr, reward_attr, metric, mode, perturbation_interval,
+            hyperparam_mutations, quantile_fraction, resample_probability,
+            custom_explore_fn, log_config, require_attrs, synch)
+
+        self.last_exploration_time = 0  # when we last explored
+        self.data = pd.DataFrame()
+
+        self.hyperparam_bounds = hyperparam_bounds
+
+        # Current = trials running that have already re-started after reaching
+        #           the checkpoint. When exploring we care if these trials
+        #           are already in or scheduled to be in the next round. 
+        self.current = None 
+
+    def _save_trial_state(self, state, time, result, trial):
+
+        score = super(PopulationBasedTraining, self)._save_trial_state(
+            state, time, result, trial)
+
+        # Data logging for PB2.
+
+        # Collect hyperparams names and current values for this trial.
+        names = []
+        values = []
+        for key in self._hyperparam_mutations.keys():
+            names.append(str(key))
+            values.append(trial.config[key])
+
+        # Store trial state and hyperparams in dataframe.
+        # this needs to be made more general.
+        lst = [[trial, result[self._time_attr]] + values + [score]]
+        cols = ['Trial', 'Time'] + names + ['Reward']
+        entry = pd.DataFrame(lst, columns=cols)
+
+        self.data = pd.concat([self.data, entry]).reset_index(drop=True)
+        self.data.Trial = self.data.Trial.astype('str')
+
+    def _get_new_config(self, trial, trial_to_clone):
+        # If we are at a new timestep, we dont want to penalise for trials
+        # still going
+        if self.data['Time'].max() > self.last_exploration_time:
+            self.current = None
+
+        new_config, data = explore(self.data, self.hyperparam_bounds, self.current,
+                                   trial_to_clone, trial,
+                                   trial_to_clone.config)
+
+        # important to replace the old values, since we are copying across
+        self.data = data.copy()
+
+        # if the current guy youre selecting is at a point youve already done,
+        # then append the data to the "current" which is the points in the current batch
+
+        new = []
+        for key in self._hyperparam_mutations.keys():
+            new.append(new_config[key])
```

Comment:
this might be a little faster?
```suggestion
        new = [new_config[key] for key in self._hyperparam_mutations]
```

Reply:
I can't commit this since it's outdated. Can you pull the most recent changes?